### PR TITLE
UPSTREAM: 76142:  Fix TestCreateDisk by sorting tags

### PR DESCRIPTION
**Pull request by soltysh**
This should fix one of the flakes in the CI.

/assign @ingvagabund @jsafrane

### CODE REVIEW

File: vendor/k8s.io/kubernetes/pkg/cloudprovider/providers/aws/aws_test.go [hunk 1]
```diff
@@ -69,6 +70,25 @@ func (m *MockedFakeEC2) DescribeSecurityGroups(request *ec2.DescribeSecurityGrou
...
+		sort.Slice(tags, func(i, j int) bool {
+			if tags[i] == nil && tags[j] != nil {
+				return false
+			}
+			if tags[i] != nil && tags[j] == nil {
+				return true
+			}
+			return *tags[i].Key < *tags[j].Key
+		})
+	}
```
Comment (sttts) [inline vendor/k8s.io/kubernetes/pkg/cloudprovider/providers/aws/aws_test.go]: is it ok to mutate the input? looks fishy.

File: vendor/k8s.io/kubernetes/pkg/cloudprovider/providers/aws/aws_test.go [hunk 2]
```diff
@@ -69,6 +70,25 @@ func (m *MockedFakeEC2) DescribeSecurityGroups(request *ec2.DescribeSecurityGrou
 }
 
 func (m *MockedFakeEC2) CreateVolume(request *ec2.CreateVolumeInput) (*ec2.Volume, error) {
+	// mock requires stable input, and in CreateDisk we invoke buildTags which uses
+	// a map to create tags, which then get converted into an array. This leads to
+	// unstable sorting order which confuses mock. Sorted tags are not needed in
+	// regular code, but are a must in tests here:
+	for i := 0; i < len(request.TagSpecifications); i++ {
```
Comment (ingvagabund) [inline vendor/k8s.io/kubernetes/pkg/cloudprovider/providers/aws/aws_test.go]: I don't think this is enough. This is just basically sorting the tags in the `request`. Though, the real `CreateVolume` is still constructing the list of tags based on the order of iterating through map keys. We either need to sort the list of tags right before `CreateVolume` is called. Or, just specifiy `awsServices.ec2.(*MockedFakeEC2).On("CreateVolume", request).Return(volume, nil)` on more time just with different order of tags in `request`.

**Comment by ingvagabund** [comment]
Let's hold the PR for a bit. I don't think it's enough to sort the tags in faked `CreateVolume`.

/hold

File: vendor/k8s.io/kubernetes/pkg/cloudprovider/providers/aws/aws_test.go (hunk 2 above)
Comment (ingvagabund) [reply vendor/k8s.io/kubernetes/pkg/cloudprovider/providers/aws/aws_test.go]: Real `CreateVolume` is called with the following code:
```go
	request := &ec2.CreateVolumeInput{}
	...
	tags := volumeOptions.Tags
	tags = c.tagging.buildTags(ResourceLifecycleOwned, tags)

	var tagList []*ec2.Tag
	for k, v := range tags {
		tagList = append(tagList, &ec2.Tag{
			Key: aws.String(k), Value: aws.String(v),
		})
	}
	request.TagSpecifications = append(request.TagSpecifications, &ec2.TagSpecification{
		Tags:         tagList,
		ResourceType: aws.String(ec2.ResourceTypeVolume),
	})

	response, err := c.ec2.CreateVolume(request)
```

Ordering of tags in the list (i.e. `tagList`) is still constructed randomly based on the order in which `tags` (map[string]string) is iterated. Signature of `buildTags` method: `buildTags(lifecycle ResourceLifecycle, additionalTags map[string]string) map[string]string`.

**Comment by sttts** [comment]
> Let's hold the PR for a bit.

until when?

**Comment by ingvagabund** [comment]
> until when?

Until the PR gets updated. The current fix does not fix the issue. I talked to @soltysh off the PR.

**Comment by smarterclayton** [comment]
If you can't fix this disable the test and then come back.

File: vendor/k8s.io/kubernetes/pkg/cloudprovider/providers/aws/aws_test.go (hunk 2 above)
Comment (soltysh) [reply vendor/k8s.io/kubernetes/pkg/cloudprovider/providers/aws/aws_test.go]: Discussed this on slack, this is working as expected, iow. before above `CreateVolume` is being called we are sorting the tags to have stable input which match those that are in `TestCreateDisk`. I've added additional comment in `TestCreateDisk` next to `Tags` to ensure we have those values sorted.

**Comment by smarterclayton** [comment]
Force merging to help unblock the queues.

**Pull request merged**